File: rs/earthbucks_lib/src/iso_buf_reader.rs
use crate::ebx_error::EbxError;
use byteorder::{BigEndian, ReadBytesExt};
use std::io::Cursor;

pub struct IsoBufReader {
    buf: Cursor<Vec<u8>>,
}

impl IsoBufReader {
    pub fn new(buf: Vec<u8>) -> IsoBufReader {
        IsoBufReader {
            buf: Cursor::new(buf),
        }
    }

    pub fn eof(&self) -> bool {
        self.buf.position() as usize >= self.buf.get_ref().len()
    }

    pub fn remainder_len(&self) -> usize {
        self.buf.get_ref().len() - self.buf.position() as usize
    }

    pub fn read(&mut self, len: usize) -> Result<Vec<u8>, EbxError> {
        let pos = self.buf.position() as usize;
        if pos + len > self.buf.get_ref().len() {
            return Err(EbxError::NotEnoughDataError { source: None });
        }
        let buf = self.buf.get_ref()[pos..pos + len].to_vec();
        self.buf.set_position((pos + len) as u64);
        Ok(buf)
    }

    pub fn read_remainder(&mut self) -> Vec<u8> {
        let pos = self.buf.position() as usize;
        let buf = self.buf.get_ref()[pos..].to_vec();
        self.buf.set_position(self.buf.get_ref().len() as u64);
        buf
    }

    pub fn read_u8(&mut self) -> Result<u8, EbxError> {
        self.buf
            .read_u8()
            .map_err(|_| EbxError::NotEnoughDataError { source: None })
    }

    pub fn read_u16_be(&mut self) -> Result<u16, EbxError> {
        self.buf
            .read_u16::<BigEndian>()
            .map_err(|_| EbxError::NotEnoughDataError { source: None })
    }

    pub fn read_u32_be(&mut self) -> Result<u32, EbxError> {
        self.buf
            .read_u32::<BigEndian>()
            .map_err(|_| EbxError::NotEnoughDataError { source: None })
    }

    pub fn read_u64_be(&mut self) -> Result<u64, EbxError> {
        self.buf
            .read_u64::<BigEndian>()
            .map_err(|_| EbxError::NotEnoughDataError { source: None })
    }

    pub fn read_var_int_buf(&mut self) -> Result<Vec<u8>, EbxError> {
        let first = self.read_u8().map_err(|e| EbxError::NotEnoughDataError {
            source: Some(Box::new(e)),
        })?;
        match first {
            0xfd => {
                let mut buf = vec![first];
                buf.extend_from_slice(&self.read(2).map_err(|e| EbxError::NotEnoughDataError {
                    source: Some(Box::new(e)),
                })?);
                if Cursor::new(&buf[1..]).read_u16::<BigEndian>().unwrap() < 0xfd {
                    return Err(EbxError::NonMinimalEncodingError { source: None });
                }
                Ok(buf)
            }
            0xfe => {
                let mut buf = vec![first];
                buf.extend_from_slice(&self.read(4).map_err(|e| EbxError::NotEnoughDataError {
                    source: Some(Box::new(e)),
                })?);

                if Cursor::new(&buf[1..]).read_u32::<BigEndian>().unwrap() < 0x10000 {
                    return Err(EbxError::NonMinimalEncodingError { source: None });
                }
                Ok(buf)
            }
            0xff => {
                let mut buf = vec![first];
                buf.extend_from_slice(&self.read(8).map_err(|e| EbxError::NotEnoughDataError {
                    source: Some(Box::new(e)),
                })?);
                if Cursor::new(&buf[1..]).read_u64::<BigEndian>().unwrap() < 0x100000000 {
                    return Err(EbxError::NonMinimalEncodingError { source: None });
                }
                Ok(buf)
            }
            _ => Ok(vec![first]),
        }
    }

    pub fn read_var_int(&mut self) -> Result<u64, EbxError> {
        let buf = self.read_var_int_buf()?;
        let first = buf[0];
        match first {
            0xfd => Ok(Cursor::new(&buf[1..]).read_u16::<BigEndian>().unwrap() as u64),
            0xfe => Ok(Cursor::new(&buf[1..]).read_u32::<BigEndian>().unwrap() as u64),
            0xff => Ok(Cursor::new(&buf[1..]).read_u64::<BigEndian>().unwrap()),
            _ => Ok(first as u64),
        }
    }
}

File: ts/earthbucks-lib/src/iso-buf-reader.ts
import { SysBuf, FixedIsoBuf } from "./iso-buf.js";
import { Result, Ok, Err } from "earthbucks-opt-res/src/lib.js";
import {
  EbxError,
  NotEnoughDataError,
  NonMinimalEncodingError,
  InsufficientPrecisionError,
} from "./ebx-error.js";
import { U8, U16, U32, U64 } from "./numbers.js";

export class IsoBufReader {
  buf: SysBuf;
  pos: number;

  constructor(buf: SysBuf) {
    this.buf = buf;
    this.pos = 0;
  }

  eof(): boolean {
    return this.pos >= this.buf.length;
  }

  read(len: number): Result<SysBuf, EbxError> {
    if (this.pos + len > this.buf.length) {
      return Err(new NotEnoughDataError());
    }
    const buf = this.buf.subarray(this.pos, this.pos + len);
    const newBuf = SysBuf.alloc(len);
    newBuf.set(buf);
    this.pos += len;
    return Ok(newBuf);
  }

  readFixed<N extends number>(len: N): Result<FixedIsoBuf<N>, EbxError> {
    const res = this.read(len);
    if (res.err) {
      return Err(res.val);
    }
    return FixedIsoBuf.fromBuf(len, res.unwrap()) as Result<
      FixedIsoBuf<N>,
      EbxError
    >;
  }

  readRemainder(): SysBuf {
    return this.read(this.buf.length - this.pos).unwrap();
  }

  readU8(): Result<U8, EbxError> {
    let val: number;
    try {
      val = this.buf.readUInt8(this.pos);
    } catch (err: unknown) {
      return Err(new NotEnoughDataError(err as Error));
    }
    this.pos += 1;
    return Ok(new U8(BigInt(val)));
  }

  readU16BE(): Result<U16, EbxError> {
    let val: number;
    try {
      val = this.buf.readUInt16BE(this.pos);
    } catch (err) {
      return Err(new NotEnoughDataError(err as Error));
    }
    this.pos += 2;
    return Ok(new U16(BigInt(val)));
  }

  readU32BE(): Result<U32, EbxError> {
    let val: number;
    try {
      val = this.buf.readUInt32BE(this.pos);
    } catch (err) {
      return Err(new NotEnoughDataError(err as Error));
    }
    this.pos += 4;
    return Ok(new U32(BigInt(val)));
  }

  readU64BE(): Result<U64, EbxError> {
    let val: bigint;
    try {
      val = this.buf.readBigUInt64BE(this.pos);
    } catch (err) {
      return Err(new NotEnoughDataError(err as Error));
    }
    this.pos += 8;
    return Ok(new U64(val));
  }

  readVarIntBuf(): Result<SysBuf, EbxError> {
    const res = this.readU8();
    if (res.err) {
      return Err(new NotEnoughDataError(res.val));
    }
    const first = res.unwrap().n;
    if (first === 0xfd) {
      const res = this.read(2);
      if (res.err) {
        return Err(new NotEnoughDataError(res.val));
      }
      const buf = res.unwrap();
      if (buf.readUInt16BE(0) < 0xfd) {
        return Err(new NonMinimalEncodingError());
      }
      return Ok(SysBuf.concat([SysBuf.from([first]), buf]));
    } else if (first === 0xfe) {
      const res = this.read(4);
      if (res.err) {
        return Err(new NotEnoughDataError(res.val));
      }
      const buf = res.unwrap();
      if (buf.readUInt32BE(0) < 0x10000) {
        return Err(new NonMinimalEncodingError());
      }
      return Ok(SysBuf.concat([SysBuf.from([first]), buf]));
    } else if (first === 0xff) {
      const res = this.read(8);
      if (res.err) {
        return Err(new NotEnoughDataError(res.val));
      }
      const buf = res.unwrap();
      const bn = buf.readBigUInt64BE(0);
      if (bn < 0x100000000) {
        return Err(new NonMinimalEncodingError());
      }
      return Ok(SysBuf.concat([SysBuf.from([first]), buf]));
    } else {
      return Ok(SysBuf.from([first]));
    }
  }

  readVarInt(): Result<U64, EbxError> {
    const res = this.readVarIntBuf();
    if (res.err) {
      return Err(res.val);
    }
    const buf = res.unwrap();
    const first = buf.readUInt8(0);
    let value: bigint;
    switch (first) {
      case 0xfd:
        value = BigInt(buf.readUInt16BE(1));
        break;
      case 0xfe:
        value = BigInt(buf.readUInt32BE(1));
        break;
      case 0xff:
        value = buf.readBigUInt64BE(1);
        break;
      default:
        value = BigInt(first);
        break;
    }
    return Ok(new U64(value));
  }
}
